deepblast/nw_cuda.py

```diff
@@ -0,0 +1,314 @@
+            raise TypeError("CUDA variant only supports torch.float32 type")
+
+        # Return both the alignment matrix
+        B, N, M = theta.shape
+
+        Q = torch.zeros((B, N + 2, M + 2, 3), dtype=theta.dtype)
+        Vt = torch.zeros((B), dtype=theta.dtype)
+        bpg = (B + (tpb - 1)) // tpb  # blocks per grid
+
+        d_theta = cuda.to_device(theta.detach().numpy())
+        d_A = cuda.to_device(A.detach().numpy())
+        d_Q = cuda.to_device(Q.detach().numpy())
+        d_Vt = cuda.to_device(Vt.detach().numpy())
+
+        _forward_pass_kernel[tpb, bpg](d_theta, d_A, d_Q, d_Vt)
+        d_Vt.copy_to_host(Vt.detach().numpy())
+
+        ctx.save_for_backward(theta, A, Q)
+        ctx.others = operator
+        ctx.device_arrays = d_Q
+        return Vt
+
+    @staticmethod
+    def backward(ctx, Et):
+        """
+        Parameters
+        ----------
+        ctx : ?
+           Some autograd context object
+        Et : torch.Tensor
+           Last alignment trace (scalar value)
+        """
+        theta, A, Q = ctx.saved_tensors
+        d_Q = ctx.device_arrays
+        operator = ctx.others
+
+        E, A = NeedlemanWunschFunctionBackward.apply(theta, A, Et, d_Q, operator)
+        return E[:, 1:-1, 1:-1], A, None, None, None
+
+
+class NeedlemanWunschFunctionBackward(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, theta, A, Et, d_Q, operator):
+        if operator != 'softmax':
+            raise NotImplementedError(
+                "CUDA variant only supports 'softmax' operator")
+
+        B, N, M = theta.shape
+
+        E = torch.zeros((B, N + 2, M + 2), dtype=theta.dtype)
+
+        d_Et = cuda.to_device(Et.detach().numpy())
+        d_E = cuda.to_device(E.detach().numpy())
+        bpg = (B + (tpb - 1)) // tpb  # blocks per grid
+
+        _backward_pass_kernel[tpb, bpg](d_Et, d_Q, d_E)
+        d_E.copy_to_host(E.detach().numpy())
+
+        ctx.others = operator
+        ctx.device_arrays = d_Q, d_E
+        return E, A
+
+    @staticmethod
+    def backward(ctx, Ztheta, ZA):
+        """
+        Parameters
+        ----------
+        ctx : ?
+           Some autograd context object
+        Ztheta : torch.Tensor
+            Derivative of theta of dimension N x M
+        ZA : torch.Tensor
+            Derivative of transition matrix of dimension 3 x 3
+        """
+        # operator = ctx.others
+        d_Q, d_E  = ctx.device_arrays
+        
+        B, ZN, ZM = Ztheta.shape
+        bpg = (B + (tpb - 1)) // tpb  # blocks per grid
+
+        Qd = torch.zeros((B, ZN, ZM, 3), dtype=Ztheta.dtype)
+        Vtd = torch.zeros(B, dtype=Ztheta.dtype)
+        Ed = torch.zeros((B, ZN, ZM), dtype=Ztheta.dtype)
+
+        d_Ztheta = cuda.to_device(Ztheta.detach().numpy())
+        d_ZA = cuda.to_device(ZA.detach().numpy())
+        d_Vtd = cuda.to_device(Vtd.detach().numpy())
+        d_Qd = cuda.to_device(Qd.detach().numpy())
+        d_Ed = cuda.to_device(Ed.detach().numpy())
```

Comment:
Apparently in order to call numpy(), the data needs to be first be moved to the CPU.
I wonder if there is a way to bypass this (i.e. avoid calling numpy() in order to keep the data on the GPU).